### PR TITLE
statistics: record last gc ts to avoid huge read on stats table

### LINKED ISSUE #31778: GCStats runs too many times
## Enhancement
There are ~4000 tables in tidb-server, and only 78 of them are frequently accessed.

From the dashboard we can see that the SQLs triggered by GCStats run too frequently.
截屏2022-01-18 下午9 22 22

Which causes the CPU to oscillate every 5mins:
截屏2022-01-18 下午9 24 21

### LINKED ISSUE #45966: Huge memory pulse in non foreground workload cluster
## Enhancement
image
We can observe memory pulse periodically.

It's possible that the statistics' async GC causes this.
The following profile is the CPU profile when the memory goes high.
45bbca2f-f52c-4960-a04f-ee5a23124cf2

526e4399-c798-4102-a3ba-6aec34d31d08

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/45966, close #31778 

Problem Summary:

The previous GCStats will read nearly all the stats into the memory

### What is changed and how it works?

This pull tries to limit its influence.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: statistics/handle/ddl.go [hunk 1]
```diff
@@ -320,6 +332,36 @@ func (h *Handle) insertTableStats2KV(info *model.TableInfo, physicalID int64) (e
...
 
+// resetTableStats2KV resets the count to 0.
+func (h *Handle) resetTableStats2KVForDrop(physicalID int64) (err error) {
+	statsVer := uint64(0)
+	defer func() {
+		if err == nil && statsVer != 0 {
+			h.recordHistoricalStatsMeta(physicalID, statsVer, StatsMetaHistorySourceSchemaChange)
+		}
+	}()
+	h.mu.Lock()
```
Comment (AilinKid) [inline statistics/handle/ddl.go:343]: can we use the h.session.pool to avoid the lock the handle?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: session/bootstrap.go [hunk 2]
```diff
@@ -728,6 +728,9 @@ const (
 	// The variable name in mysql.tidb table and it records the default value of
 	// oom-action when upgrade from v3.0.x to v4.0.11+.
 	tidbDefOOMAction = "default_oom_action"
+
+	tiDBStatsGCLastTS        = "tidb_stats_gc_last_ts"
```
Comment (time-and-fate) [inline session/bootstrap.go:732]: We defined this const string twice (another in `statistics/handle/gc.go`), we can merge them to make it more clear when other people are trying to find their usages.

File: session/bootstrap.go [hunk 3]
```diff
@@ -2721,6 +2729,18 @@ func upgradeToVer173(s Session, ver int64) {
...
+	if ver >= version174 {
+		return
+	}
+	writeStatsGCLastPos(s)
+}
+
+func writeStatsGCLastPos(s Session) {
+	mustExecute(s, "INSERT HIGH_PRIORITY INTO %n.%n VALUES(%?, %?, %?) ON DUPLICATE KEY UPDATE VARIABLE_VALUE = %?",
+		mysql.SystemDB, mysql.TiDBTable, tiDBStatsGCLastTS, 0, tiDBStatsGCLastTSComment, 0)
+}
```
Comment (time-and-fate) [inline session/bootstrap.go:2742]: I think we don't need this actually.
We can just treat the empty result as `0` when reading this record, and use `insert ... on duplicate key update` when writing this record.

File: statistics/handle/ddl.go [hunk 4]
```diff
@@ -320,6 +332,36 @@ func (h *Handle) insertTableStats2KV(info *model.TableInfo, physicalID int64) (e
...
+	}
+	defer func() {
+		err = finishTransaction(ctx, exec, err)
+	}()
+	txn, err := h.mu.ctx.Txn(true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	startTS := txn.StartTS()
+	if _, err := exec.ExecuteInternal(ctx, "update mysql.stats_meta set count=0, modify_count=0, version=%? where table_id =%?", startTS, physicalID); err != nil {
```
Comment (time-and-fate) [inline statistics/handle/ddl.go:359]: Why are we resetting the row count to `0`?
I think updating the version is enough.
Here we have not deleted the actual stats yet. Resetting the row count will create an inconsistency between the `stats_meta` and actual stats, which is unnecessary in my opinion.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**